### ISSUE TITLE
unknown free variable: _kernel_fresh.1210

### ISSUE THREAD

**Issue by nomeata** [leanprover/lean4#4548]
### Prerequisites

I stumbled upon this while working on lean, this is what I minimized it to so far:

```
import Lean

open Lean Meta

opaque foo : MetaM Unit

-- unknown free variable: _kernel_fresh.1210
partial def bar : MetaM Unit := do -- doesn't happen with IO here
  let nonIndices : Array Unit := #[].filter (fun i => true) -- doesn't happen without filter here
  for i in nonIndices do
    try
      foo -- needs this foo, pure () is not enough
      return
    catch e => pure () -- needs pure here, return doesnt work
```

Could be a duplicate of #1415 or #4418, but it is hard to tell without a full investigation, so reporting separately to avoid confusion if a fix ends up fixing one some of these instances.

### Versions

4.9.0-rc2, 4.10.0-nightly-2024-06-23,
[Output of `#eval Lean.versionString`]
[OS version, if not using live.lean-lang.org.]

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others are impacted by this issue, please ask them to add :+1: to it.

**Comment by nomeata** [comment]
Ah, adding some mostly unrelated lines to my function makes the error go away. So I'm at least unblocked.

**Comment by leodemoura** [comment]
We can confirm that is a code generator issue by using `noncomputable`

**Comment by JLimperg** [comment]
Probably another instance of this bug, minimised from some Aesop code:

```lean
import Lean

open Lean Lean.Parser.Tactic

/-
unknown free variable: _kernel_fresh.134
-/
def foo {m} [Monad m] [MonadQuotation m] (tacs : Array (TSyntax `tactic)) (ns : Array Ident) (b : Bool) : m Syntax := do
  let usedNames : HashSet Name := tacs[1:].foldl (init := sorry) sorry
  for n in ns do
    if usedNames.contains sorry then
      continue
  if b then
    let tacs : TSyntax ``tacticSeq ← `(tacticSeq| $tacs:tactic*)
    sorry
  else
    sorry
```

**Comment by dupuisf** [comment]
I was told by @TwoFX that this might be another instance of this bug:
```lean
import Std.Data.HashSet.Basic

def validUpdate (upd : Array Nat) (rules : Std.HashSet (Nat × Nat)) : Bool := Id.run do
  for i in [:upd.size-1] do
    for hj : j in [i+1:upd.size] do
      if rules.contains ⟨upd[j], upd[i]!⟩ then return false
  return true
```

The resulting  error here is `unknown free variable: _kernel_fresh.278`, which disappears if one replaces `Std.HashSet` with `Array`.

**Comment by marcusrossel** [comment]
Here's another one:

```
import Lean

#version -- Lean 4.15.0

example : Std.HashSet Nat := Id.run do
  if "".isEmpty then dbg_trace ""; return ∅
  if (#[] : Array Nat).all (fun _ => True) then return ∅
  return ∅
```

**Comment by zwarich** [comment]
All of these are fixed by the new compiler.
